### PR TITLE
Backport: services via gateway, minus API changes

### PR DESCRIPTION
**Pull request by vito**
This is a halfway step to #5557 that hoists over its internal changes while keeping the existing API unchanged.

### CODE REVIEW

**Comment by vito** [comment]
@marcosnils @sipsma Unfortunately this doesn't resolve the Cloud issue with services showing as failed. Now [they show as running](https://alpha.dagger.cloud/runs/912d156d-4a35-4bec-b5be-fc00fa9dd75d), which is maybe 1% better and even technically correct (explained below), but it's still not what we want. I have an idea in mind but it'll require a small tweak to every SDK.

It doesn't work because services stop asynchronously. The client disconnects, which will stop them all via plain old `ctx` cancellation, but since the client has already disconnected it can't receive any more progress. So the client can never see a `Completed` status for services as long as their stopping is triggered by disconnecting.

What I have in mind is to add some kind of `Query { close: Void! }` (strawman) API that clients call just before closing, which will release all of the client's background resources and wait. In this case, that would mean stopping all services and emitting `Completed` statuses along the way. Any thoughts (including "ew no")?

It's a bit unusual, and is another subtle thing each SDK would have to remember to implement. Are there any other scenarios where this API would come in useful?

Another option could be to implement some sort of internal shutdown protocol, but there'd still need to be a clear and obvious way for that to be invoked by clients backed by all connection types.

Yet another option could be to leave this as-is but add some sort of metadata on the vertex that Cloud detects in order to recognize "oh, this is a service, so I know that it was just left running in the background." The upside here is that SDKs don't have to change, and closing can still be async, likely allowing for snappier DX. Fixing this in Cloud kind of defeats the purpose of this PR though, since a similar workaround could probably be implemented without it, and these changes are already included in #5557.

**Comment by vito** [comment]
Went with above API for max async, but still open to other ideas. Just want to have something ready to release if we want it. Marked the API "Experimental" so we can change or remove it whenever. Took a swing at Go, Elixir (@wingyplus :bow:), and NodeJS implementations. Could use help with Rust and Python.

**Comment by vito** [comment]
Confirmed that the above change resolves the issue: [this build](https://alpha.dagger.cloud/runs/aa907c90-773c-4b59-b837-8981e03f2e8b) shows its two services as completed. The user code just needs to close the client, which it should be doing already, and the SDK's `close()` needs to call `Stop`.

I went with a configurable timeout just to make it easier for SDKs to implement. I've set it to 10 seconds in the SDKs I've updated so far.

File: sdk/elixir/lib/dagger.ex
Comment (wingyplus) [inline sdk/elixir/lib/dagger.ex:47]: The timeout is quite weird for me when first read it. Because in Erlang/Elixir, they use milliseconds. In my head thought, why timeout is super quick? 😂

Comment (wingyplus) [reply sdk/elixir/lib/dagger.ex:47]: That means we need timeout at the request scope right? If that is so, I think it would be easy but need to bypass timeout to the HTTP client that uses it internally.

Comment (vito) [reply sdk/elixir/lib/dagger.ex:47]: Yeah, timing out the request would be sufficient. It's technically different from now, where the server implements a timeout internal to the request and will complete the request, but I think it'd be functionally equivalent for all intents and purposes.

**Comment by vito** [comment]
SDK implementation status for the above `Query { stop }` API:

* [x] Go
* [x] NodeJS
* [x] Elixir
* [x] Python
* [ ] Rust
* [ ] Dotnet
* [ ] Java?

Really just guessing my way through most of these to get a head start. Requesting reviews now.

**Comment by helderco** [comment]
Sorry @vito, I should have given an heads up that I spent a bit of time today taking care of the Python bit. I'll replace it in a minute.

**Comment by vito** [comment]
@helderco All good, feel free to replace it! :pray:

**Comment by gerhard** [comment]
@vito re various SDKs:
- `Dotnet` is the most experimental one, nothing to do there
- `Java` is yet to be merged - https://github.com/dagger/dagger/pull/5504 - nothing to do there (blocked on me)
- `Rust` is something that @kjuulh can help with post-merge, no need to block

**Comment by helderco** [comment]
The reverts were from a GUI, weren't signed-off, I'm going to force push to fix.

**Comment by kjuulh** [comment]
Yep, I will handle the follow up inconsistencies for rust, if the tests don't complain that is.

**Comment by helderco** [comment]
@vito what should be the behavior in `dagger run` and in Zenith? Should the SDK call stop in both cases?

**Comment by vito** [comment]
@helderco That's a good question. It seems that we should only call `stop` when we know the `ClientID` is going away for good. Otherwise you might end up stopping the services started by a separate client in the same session, before they're actually done being used. For example, you might `dagger run` a script that spawns a bunch of pipelines in parallel, each with their own client. If we naively call `stop` in this case, one concurrent client would stop all the other client's services.

So with that rule, it seems most correct to _not_ call `stop` in the `DAGGER_SESSION_PORT` case, because you can't be sure that the session is completed.

For Zenith/Dagger-in-Dagger, we can work with the above constraint by having it automatically do the equivalent of the `stop` API in the same place where the session is created (the shim).

In that case I think the current Go, NodeJS, and Elixir SDKs implementations are not right as they will unconditionally stop no matter what the internal connection type is.

Thinking about it more, if we're now saying this only needs to happen only in the case where a `dagger session` is spawned, I wonder if we can just handle it there, without adding a whole new API? Basically the client closing would send a signal to the process somehow, which would trigger all the stopping, and we'd wait for that to complete, which should be roughly already what's happening. I'm pretty sure I looked into this already, but I don't remember why I backed out. :sweat_smile: Looking into it now.

File: core/services.go
Comment (sipsma) [inline core/services.go:240]: Possible synchronization bug:
1. Multiple goroutines are blocked on `starting.Wait()` on line 232
2. Starting fails, so those goroutines all unblock and continue to `svc.Start` on line 250
3. I think that may already be unexpected behavior, but also `ss.bindings` will always be set to `1` on 262

Not sure if I'm just misunderstanding since it's all pretty tricky obviously.

File: core/services.go
Comment (sipsma) [inline core/services.go:294]:
```suggestion
		// already running; stop it
```

File: core/services.go
Comment (sipsma) [inline core/services.go:304]:
```suggestion
			// starting succeeded as normal; stop the instance
```

**Review by sipsma — COMMENTED** [review]
Left some comments, saw after I started reviewing that you're making a few more updates so apologies if a few of the comments are more nit-picky than is helpful now :-)

**Comment by vito** [comment]
Pushed a prospective switch to a new internal HTTP endpoint, `/shutdown`, which gets called on `(engine/client.Client).Close`. I'm on shoddy internet, so relying on CI to tell me if it works. :sweat_smile: I reverted all the `sdk/` changes, but they're still in the history if you want to salvage any part of your changes @helderco (it looked like part of that was refactoring)

**Comment by helderco** [comment]
Yeah, I'll keep all of it except for just the stop call but I'll do it in another PR. 😁

File: core/services.go
Comment (vito) [reply core/services.go]: Yeah this was definitely sus. Pushed a theoretical fix, but I'll admit I haven't reproduced this first. Unit tests are in order, if I can isolate the concurrency aspect. :thinking:

File: core/net.go
Comment (vito) [reply core/net.go]: I can actually just remove this since it's part of the h2c/c2h stuff, but I'll address this in #5557!

**Review by TomChv — COMMENTED** [review]
Changes looks, I need to try on my local host to verify if everything is okay.

Will the SDK clients have an update? I saw the old implementation and see that we now use `shutdown` but as far as I understand, this is supported to be called by the client?

Do I need to wait before my attempt to test the PR?

**Comment by vito** [comment]
@TomChv Thanks for taking a look! SDKs no longer need updating - I'll minimize those comments to avoid confusion.

This PR is ready for review + testing, but I'm going to add some unit tests for the extra-tricky service lifecycle synchronization since @sipsma noticed some bugs there. They're theoretically fixed now, but I would like to have higher confidence in it.

File: sdk/elixir/lib/dagger.ex
Comment (vito) [reply sdk/elixir/lib/dagger.ex]: @wingyplus Just following up before I mark this resolved in case you started on anything here: SDKs no longer need to be updated. :tada:

**Comment by vito** [comment]
Fully ready now! Backfilled unit tests for the tricky service lifecycle and did some follow-up cleanups.

File: engine/client/client.go
Comment (wingyplus) [inline engine/client/client.go:308]: This solution is great!!! We don't need to deal with each SDK now. 👍

**Comment by marcosnils** [comment]
LGTM from the integration testing perspective.

**Comment by jpadams** [comment]
I tested this on Go, Python, Node.js SDKs using the `examples/sdk/*/db-service` example with an engine from this PR brought up with `./hack/dev` compared to the `v0.8.4` engine.

In each case I got the old undesirable "running" behavior from `v0.8.4` and the desired behavior from this PR build.

I ran some unrelated things for good measure. Looks good!

Node.js
image

Golang
image

Python
image

**Review by jpadams — APPROVED** [review]
From my perspective of testing this live, I've had no issues with this at all. Seems super solid to me.

**Comment by jpadams** [comment]
@TomChv any blocker to you giving this the 👍

**Review by gerhard — APPROVED** [review]
This was a pleasure to read! Straightforward code, good tests & some nice refactorings along the way. While we don't have the new API yet - I was especially excited about not starting services when ops that require them are cached - everything else works as expected.

Before this PR, services wouldn't show as stopped in the interactive TUI (@jpadams mentioned it in the context of Cloud):

https://github.com/dagger/dagger/assets/3342/9eff2f26-8b35-4ace-ae44-d4771b6a3d68

With the changes in this PR, services show as stopped:

https://github.com/dagger/dagger/assets/3342/98149adb-9503-4a42-be47-9cca9ad9bfde

Same command with `dagger run --debug`:

https://github.com/dagger/dagger/assets/3342/013aea46-2851-47bb-a767-86ce47ad7805

@vito what would you like to add to the changelog, via `changie new`?

**Pull request merged**
